### PR TITLE
Add theme colors to swatch color picker

### PR DESCRIPTION
**Pull request by zoarif**
#### Pull request checklist
Swatch color picker does not support theme colors yet. We need to add them since we are using the color picker in RTE property pane.

#### Description of changes
I have added theme classes that get added to the color if a theme color is passed in as prop. 

#### Focus areas to test
Tested and added examples on example page

### CODE REVIEW

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.Props.ts [hunk 1]
```diff
@@ -76,6 +76,32 @@ export interface ISwatchColorPickerProps {
   shouldFocusCircularNavigate?: boolean;
 }
 
+/**
+ * Represents theme colors
+ */
+export enum SwatchThemeColor {
```
Comment (antonlabunets) [inline packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.Props.ts]: Should this be a `const enum`?

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.tsx [hunk 2]
```diff
@@ -176,9 +177,22 @@ export class SwatchColorPicker extends BaseComponent<ISwatchColorPickerProps, IS
...
+        themedClass
+      );
+
     // Build an SVG for the cell with the given shape and color properties
     return (
-      <svg className={ css(styles.svg, this.props.cellShape, this.props.cellShape === 'circle' ? styles.circle : '') } viewBox='0 0 20 20' fill={ getColorFromString(colorOption.color as string).str } >
+      <svg
+        className={ svgClassName }
+        viewBox='0 0 20 20'
+        fill={ themedClass === '' ? getColorFromString(colorOption.color as string).str : undefined }
```
Comment (antonlabunets) [inline packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.tsx]: what if it's undefined?

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.tsx [hunk 3]
```diff
@@ -189,6 +203,57 @@ export class SwatchColorPicker extends BaseComponent<ISwatchColorPickerProps, IS
...
+      case SwatchThemeColor.ThemeLighterAlt:
+        return styles.themeLighterAlt;
+      case SwatchThemeColor.ThemePrimary:
+        return styles.themePrimary;
+      case SwatchThemeColor.ThemeSecondary:
+        return styles.themeSecondary;
+      case SwatchThemeColor.ThemeTertiary:
+        return styles.themeTertiary;
+      default:
+        return '';
```
Comment (antonlabunets) [inline packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.tsx]: Why not `undefined`?

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/examples/SwatchColorPicker.Basic.Example.tsx [hunk 4]
```diff
@@ -45,6 +45,19 @@ export class SwatchColorPickerBasicExample extends React.Component<any, IBasicSw
             ]
           }
         />
+        <div>Simple square swatch color picker with theme colors:</div>
+        <SwatchColorPicker
+          columnCount={ 4 }
+          cellShape={ 'square' }
+          colorCells={
+            [
+              { id: 'a', label: 'green', color: SwatchThemeColor.ThemePrimary },
```
Comment (antonlabunets) [inline packages/office-ui-fabric-react/src/components/SwatchColorPicker/examples/SwatchColorPicker.Basic.Example.tsx]: Why is the label green? Are you sure that's the primary? Why not use `primary`

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.scss [hunk 5]
```diff
@@ -125,4 +125,82 @@
   min-width: 180px;
 }
 
+.neutralDark {
```
Comment (jspurlin) [inline packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.scss]: This components should be glamorized. Also, it's unclear to me why the creator of a SwatchColorPicker wouldn't just use theme.palette.<name> (which returns a string) when building the component instead of having to redefine the themeColors

**Comment by christiango** [comment]
This is adding quite a bit of bloat to the component for existing consumers.

**Comment by christiango** [comment]
I don't think we want to bake in the concept of a theme into this component. Instead, the caller should just be passing in a list of colors (which may have come from the theme). I'm not sure I follow the scenario here.

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/examples/SwatchColorPicker.Basic.Example.tsx (hunk 4 above)
Comment (zoarif) [reply packages/office-ui-fabric-react/src/components/SwatchColorPicker/examples/SwatchColorPicker.Basic.Example.tsx]: left by mistake

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.scss (hunk 5 above)
Comment (zoarif) [reply packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.scss]: The problem with that is how would you know the color belongs to theme, that would just apply the color and won't be themed. Which is something that we don't want. Or is there a way of translating it to theme color that I am not aware of?

**Comment by zoarif** [comment]
@christiango, the scenario is this. We want that user should be able to pass in theme colors which means that when we change the theme, the colors would change with theme as well. That is why we cannot just pass in color hash,

**Comment by christiango** [comment]
Theme handling shouldn't be something that the swatch picker does though. It should work on color codes directly. I don't fully understand what you are trying to do, but it seems like you would define the list of colors in some wrapper component as the theme colors. Whenever the theme changes, you just need to make sure that the wrapper component rerenders and the list of theme colors would be updated to reflect the new theme. 

In other words, I think the issue you are running into is that a theme change isn't triggering your component to rerender of the component that defines the colors for the swatch picker.

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.scss (hunk 5 above)
Comment (christiango) [reply packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.scss]: The component shouldn't know about themes or a specific set of colors. The colors are passed in as props to the component. Those colors could happen to come from the theme, but the component doesn't need to do that. Your app would just need to make sure that the colors passed into the swatch picker are updated when the theme changes.

**Comment by zoarif** [comment]
@christiango , yes you are right about the scenario I am running into, Now I have removed the logic of using extra enums. Now I am just filling in the color based on the theme color passed in.

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.scss (hunk 5 above)
Comment (zoarif) [reply packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.scss]: I have updated the code. Please take a look

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.classNames.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+      {
+        fill: themedColor
+      } :
+      undefined
+    ]
+  });
+
+});
+
+function getSwatchThemeColor(theme: ITheme, color: string) {
```
Comment (christiango) [inline packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.classNames.ts]: Again, the color picker doesn't need to know about the theme. Your app should just be passing in color codes. When the theme changes, just rerender the component with the new color values.

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.Props.ts [hunk 7]
```diff
@@ -74,6 +75,11 @@ export interface ISwatchColorPickerProps {
    * Should focus cycle to the beginning of once the user navigates past the end (and visa vsersa). Defaults to true
    */
   shouldFocusCircularNavigate?: boolean;
+
+  /**
+ * Theme provided by HOC.
+ */
+  theme?: ITheme;
```
Comment (christiango) [inline packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.Props.ts]: I don't think this is necessary. The component itself shouldn't need to know if the colors come from a theme.

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.classNames.ts (hunk 6 above)
Comment (zoarif) [reply packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.classNames.ts]: Don't know how that works. If I do fill: color, how would that work?

Comment (christiango) [reply packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.classNames.ts]: You don't need to add any code to swatch color picker to get it to render a theme color, it just works. See this [code pen](https://codepen.io/anon/pen/xXpQrg). If you want it to update when the theme changes, just rerender the component.

Comment (zoarif) [reply packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.classNames.ts]: Then again. this is not what my issue is. I know it does but then if I change the color of my site, the primary color won't change with theme.

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.Props.ts (hunk 7 above)
Comment (zoarif) [reply packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.Props.ts]: I talked to David and he agrees that I need to use customizable for this, in order to listen to theme changes.

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.classNames.ts (hunk 6 above)
Comment (christiango) [reply packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.classNames.ts]: It will change if you rerender the component. Every time the component renders, it get the value of the theme. It seems like your problem is that your site isn't rerendering the color picker when the swatch value changes.

Comment (christiango) [reply packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.classNames.ts]: Here's an example in a [codepen](https://codepen.io/anon/pen/JrMVJZ). Clicking the button updates the theme and rerenders the swatch picker. Obviously, you wouldn't do this with a force update in real code, but it shows the idea.

File: packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.classNames.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  const palette: IPalette = theme.palette;
+  switch (color) {
```
Comment (christiango) [inline packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.classNames.ts]: How is this function any different than (color: string) => color? Not sure I follow what's going on here

Comment (christiango) [reply packages/office-ui-fabric-react/src/components/SwatchColorPicker/SwatchColorPicker.classNames.ts]: Sure it returns undefined if there isn't a color in the theme for it, but this will quickly get out of date.

**Comment by jspurlin** [comment]
In general this change doesn't make sense. All of this functionality can be achieved without touching the component. The creator of a SwatchColorPicker should be handling what colors to pass in (e.g. theme color in this example)... this will yield the same result that this PR is trying to achieve

**Comment by zoarif** [comment]
@christiango, Thanks I can understand now that if I use customizable in my code, I don't need to make a change in the Swatch color picker. Thanks for guiding me through it. Will close this pull request and update my code.
